### ISSUE TITLE
Getting HyperSpy in Debian repository

### ISSUE THREAD

**Issue by magnunor** [hyperspy/hyperspy#455]
It would be very nice having HyperSpy releases in the Debian repository, and since Ubuntu basically use the Debian repository it would be available there as well.

I'm guessing it won't be too complicated, since we already have a working Debian package. Someone just have to jump through the various Debian http://www.debian.org/doc/manuals/maint-guide/ hoops.

I could probably have a look at it, and see how it goes :)

**Comment by francisco-dlp** [comment]
This is much needed. We create the debian package using stdeb, which most likely doesn't produce a Debian compliant package. In addition to  the document that you mention it might be worth having a look at the  [Debian Python packaging guide](https://wiki.debian.org/Python/Packaging).

**Comment by magnunor** [comment]
I've had a little look into this, and Debian is planning on phasing out Python 2 in favour of Python 3.
What is the status for getting HyperSpy to run in Python 3?

**Comment by to266** [comment]
Basically `traitsui` is the only thing holding us back. We can either get rid of it (`traitsui` only) and replace it with Qt, or replace both `traits` and `traitsui` with `traitlets` and (if I understood that correctly) Qt respectively.

I don't think we started any of those steps, but I would be willing to help after the 0.9 release (which will have a major contribution from my side, so probably I will have enough things to work on)

**Comment by vidartf** [comment]
I could probably help out if we're porting to use Qt. It would be good to do that at the same time as we implement the proposed event structures (#423).

**Comment by to266** [comment]
I'm all for it. Starting next week though ;)

**Comment by magnunor** [comment]
Reviving this again.

Anything blocking this? Any dependencies which are not in the Debian repositories?

**Comment by magnunor** [comment]
I've started looking into this (again). Seeing if I can get it to build locally on my computer using the Debian Python build system.

**Comment by magnunor** [comment]
Also, getting HyperSpyUI into the Debian repositories would be nice. I'll have a look at that as well when (or if) we get HyperSpy included.

**Comment by francisco-dlp** [comment]
Did  you have success with your debian building experiment?

I think that all dependencies are in Debian. The only one that is not is dask (required for #1219) but they're currently [packaging it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=817777).

**Comment by magnunor** [comment]
I got the build system running yesterday, but was missing some packages so the building failed. I'll look into it (probably) during the weekend.

**Comment by magnunor** [comment]
Is there a python 3 version of traitsui in Debian? There is a python 2 version: https://packages.debian.org/stretch/python-traitsui

Is it rolled into the traits package? https://packages.debian.org/stretch/python3-traits ?

**Comment by francisco-dlp** [comment]
I am afraid that the answer is no :( If this is the bottleneck i am all for finally making traitsui an optional dependency.

**Comment by magnunor** [comment]
That would mean the pop-up windows when running things like `hs.load()` or `m.fit_component()` would not work with a base Debian install?

I'm a little bit surprised `python3-traitsui` is not in the Debian repos, especially since both `python-traitsui` and `python3-traits` is in there...

But yeah, as it currently stands we can't have the python3 version of `traitsui` as a dependency for HyperSpy in Debian.

We could maybe open an Issue over at https://github.com/enthought/traitsui/issues and ask if there are any plans for getting the python3 version into the Debian repos.

**Comment by francisco-dlp** [comment]
Probably we should request it/ report it to the Debian packagers instead? I think that the issue is that traitsui 5 (the one adding python3 support) is not yet in Debian.

In any case, it'll be good to make traitsui optional because another issue with traitsui is that it doesn't support qt5. For hyperspy it makes more sense to implement the ui elements using ipywidgets instead. @vidartf, how hard would it be to remove traitsui from HyperSpyUI?

**Comment by sem-geologist** [comment]
@vidartf, what about removing traitsui and replacing it with PyQt, and also moving forward from PyQt4 to PyQt5? This would no more require to downgrade anaconda, so it would please people who are using anaconda for other projects. I am already warmed up with PyQt5 (previously I was using PyQt4 extensively) and it is going anyway to be in future more preferred over Qt4 (the main issue is that Qt4 have no more Web  support, and most of distros and a lot of software is moving to Qt5). I Understand there will be a lot of work required but if it would move that direction I would be very happy to lend my hand in that. I feel very comfortable with PyQt GUI programming.
However there is some other optional issues to consider which would help to improve UI.
1. matplotlib. It is very tightly coupled with signals. Well it is one of the best library to get publication quality figures (if not the best), however for interactive work it is too laggy. It could be left as is for getting publication figures, however wouldn't pyqtgraph be better for gui? It have so much more to offer: speed, ROI's, 3D, signaling, plotting big data, extremely handy image widgets (with interactive histogram). Pyqtgraph also have more advanced docking capabilities, it would improve over-cluttered hyperspyui docking widgets (personally I cant get the console if I do not close enough of other widgets on my laptop with 1378x768 pixel screen). I think also traits could be replaced with pyqtgraph parameters and where gui is parameter tree.
2. IMHO MDI works nicely on huge screens, but on small screens it makes feel claustrophobic. Maybe it would be good idea to add option for that.

Anyway, if you would need help with GUI, I will be glad to help.

**Comment by magnunor** [comment]
Will ipywidgets work with the ipython interactive terminal as well as the notebook?

PyQtGraph looks very nice, and I guess it would be more appropriate for the interactive stuff we're doing in HyperSpy (compared to matplotlib). Getting a 3D viewer would also be very nice.

I'm guessing switching the UI stuff will be a bit of a pain? But if we're replacing traitsui anyway...

**Comment by vidartf** [comment]
Mapping out where/how traitsui is currently used:
 - gui/eels: Dialog to get convergence/collection angle + beam energy
 - gui/preferences: Preferences dialog (multi-tab built on traits available)
 - gui/messages: Utility for simple dialogs used elsewhere, e.g. OK/Cancel message boxes
 - gui/tools: A lot of different things, mostly for selecting a span in spectrum, smoothing spectrum, etc.
 - gui/eds: Dialogs for viewing/inputting EDS parameters (TEM/SEM)
 - gui/axes: Show/edit axes in an AxesManager + navigation sliders
 - gui/egerton_quatification: More 1D span selector use + dialogs
 - model + component + parameter: Dialogs for viewing/editing components. Here, the dialogs rely a lot on the automatic population from traits, but some custom code also exist.

**Comment by francisco-dlp** [comment]
It should be relatively simply to make all the ``gui/*`` functionality optional. @vidartf, how simply would it be to remove/make optional the traitsui code from component etc?

We could aim at splitting the gui code into a separate package e.g. ``hyperspy-gui-traitsui`` and  ``hyperspy-gui-ipywidgets``. We could do the same with the drawing code e.g. ``hyperspy-plotting-mpl``. In this way we could isolate the data crunching kernel which should be handy to install in headless systems.

**Comment by vidartf** [comment]
Gutting out the current code and replacing the relevant calls with exceptions is likely easy. Getting a decent system for dropping in UI front-ends might be more tricky. Not sure how that is best organized/configured.

**Comment by ericpre** [comment]
+1 for separating gui/plot features in different package, as @francisco-dlp suggested in his last comment.
Just to avoid complication in the installation process, maybe it would be good to combine this change with #1438?

**Comment by magnunor** [comment]
Now that traitsui is being made optional with https://github.com/hyperspy/hyperspy/pull/1544, and ipywidgets already being in the Debian repository (https://packages.debian.org/source/stretch/ipywidgets) packaging HyperSpy for Debian should be possible soon.

Dask is also in the Debian unstable repo: https://packages.debian.org/source/sid/dask.

I'll give it a try as soon as the ipywidgets/traitsui stuff is sorted.

**Comment by francisco-dlp** [comment]
Great!

**Comment by magnunor** [comment]
I'll have a look at this again, now that HyperSpy can be used without traitsui.

**Comment by magnunor** [comment]
I got the debian package building almost working now, but some of the tests aren't passing! Will try more in couple of days.

**Comment by francisco-dlp** [comment]
Good news!

**Comment by magnunor** [comment]
Some of the tests which are failing: https://gist.github.com/magnunor/6f5fb79ee64454993d358c5279749a3d

**Comment by ericpre** [comment]
Do you have scikit-image >= 0.13? This could explain the two first errors.

**Comment by magnunor** [comment]
Nope, seems like the most recent version in Debian unstable is 0.12.3 https://packages.debian.org/sid/python3-skimage

Ubuntu 17.10 (the current development version) has the 0.13 version

**Comment by magnunor** [comment]
`ipyparallel` and `statsmodel` are install requirements in setup.py, but I can't find Python 3 versions of them in the Debian repositories.

**Comment by magnunor** [comment]
I can now build (and install) a HyperSpy deb-package locally, however:

- I can't find `ipyparallel` in the Debian repository
- I can't find a Python 3 version of `statsmodel` in the Debian repositories
- Some of tests are not passing:
    * io/test_tiff: `AssertionError: assert <undefined> == 'µm' +  where <undefined> = <width axis, size: 1024>.units`, which is probably due to an old version of `scikit-image` in the Debian repos
    * io/test_bcf: `NameError: name 'unbcf_fast' is not defined`, which is due to me not having sorted out how to the build system to compile the cython code properly
- `python3-sympy` in the Debian repos has `texlive-fonts-extra` as a recommended package, which means the HyperSpy install needs an additional ~2 GB for all the font files. Someone already made a [thread](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=861741#5) about this in the `sympy` Debian page, and I added a comment about this in that same thread. This is not really a showstopper, but a little bit annoying.

**Comment by francisco-dlp** [comment]
Thanks for the updated. It's strange that ``statsmodels`` in missing in debian as it is a ``pandas`` requirement. We can drop the ``statsmodel`` dependency as it is only needed for the smoothers. I think that ``ipyparallel`` is not required (@to266, could you confirm this)? Regarding the skimage version, yes, hyperspy requires >=0.13 in order to support reading all the metadata from the tiff files. It should be easy to modify the tests to work with skimage < 0.13 and add a log message explaning that reading all metadata requires upgrading skimage.

**Comment by magnunor** [comment]
Sounds good! I'm guessing the best way would be to sort out all the various issues, then do either a minor or patch release (depending on what other features are in the queue).

**Comment by francisco-dlp** [comment]
:+1:

**Comment by magnunor** [comment]
Also, I think I'll first get hyperspy itself building and working properly, then package up hyperspy_gui_ipywidgets. Since hyperspy_gui_ipywidgets depends on hyperspy to work.

**Comment by ericpre** [comment]
I remember that there was a significant API change in the tifffile library, but according to [this part](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/io_plugins/tiff.py#L139) it should still be able to read the data with former scikit/tifffile versions -except for the metadata, as @francisco-dlp mentioned before.
Alternatively, would it be possible to get the last scikit-image version in the debian repository?

**Comment by to266** [comment]
Yeah, samfire can alternatively use `multiprocessing`, although `ipyparallel` is more robust.

**Comment by magnunor** [comment]
Slightly offtopic: is the functionality used by `statsmodel` and `ipyparallel` unit tested? None of the tests complained, even if these packages were not present.

**Comment by francisco-dlp** [comment]
The statsmodel features are tested in hyperspy_gui_ipywidgets solely
(something to fix).

**Comment by magnunor** [comment]
Small update:

`python3-statsmodel` is now in Debian unstable: https://packages.debian.org/unstable/python3-statsmodels

I haven't found `ipyparallel`.

`skimage` is still the old version (0.12.3).

@ericpre, it would probably be possible to get a newer `skimage` version into the Debian repos, but someone would have to spend some time to package it up.

**Comment by magnunor** [comment]
`skimage` got upgraded to 0.13.1 in Debian unstable yesterday. So the only show stopper now is `ipyparallel`.

`skimage` Debian tracking page: https://packages.qa.debian.org/s/skimage.html

**Comment by magnunor** [comment]
Major blocker is still `ipyparallel`, and I don't know if anyone is working on getting it into the Debian repositories...

How much work would there be to use `multiprocessing` (or something else) instead?

I'd really like to get HyperSpy into Debian, since that would enable me to also add `atomap` and `fpd_data_processing` as well :)

**Comment by magnunor** [comment]
Having just looked through the uses of `ipyparallel` in HyperSpy, I'm guessing it isn't really necessary to have it as dependency at all? Every place where it is used, `multiprocessing` can be used as an alternative.

If yes, could `ipyparallel` be moved to `extras_require` in `setup.py`?

**Comment by francisco-dlp** [comment]
The last time I looked into it I reached the same conclusion, although @to266 should know better, it sounds good for me.

**Comment by to266** [comment]
Yes, you can move the `ipyparallel` to extras. However if nothing's changed, the experience of using `ipyparallel` is significantly better than the `multiprocessing` module. It's been ages, though

**Comment by magnunor** [comment]
I recently went through the dependency list of HyperSpy, and the only one missing in debian unstable is `ipyparallel`. Only optional dependency missing is `mrcz`.

Not sure if I'll have time to do this myself anytime soon, but if someone else is interested...

**Comment by farhi** [comment]
We at Synchrotron Soleil have started to package HyperSpy for Debian. All dependencies are solved, and the package has been pushed to Debian NEW. The work has been subcontracted to the Freexian company, specialized in packaging.

You can follow our effort here:
- Intention to package: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=965376
- NEW report: https://ftp-master.debian.org/new/hyperspy_1.6.0-1.html

It should later appear in Debian 'SID' unstable with attached source and binary (e.g. amd64) package files.

**Comment by magnunor** [comment]
@farhi, nice! Does this also include any of the GUI packages? For example `hyperspy-gui-traitsui`?

How was ipyparallel solved? I can't see it mentioned in the Debian dependencies.

**Comment by farhi** [comment]
We are progressing slowly and iteratively. First have a core package ready, then work on adding more stuff in.

**Comment by jlaehne** [comment]
HyperSpy 1.6.1 seems to be in debian unstable and testing now, but none of the gui-packages so far:
https://packages.debian.org/sid/python3-hyperspy

**Comment by farhi** [comment]
Synchrotron Soleil has paid for the initial packaging. We are open to any further add-on, such as GIU's. Today, we favour Jupyter notebooks, so that GUI's are not our priority (so that we can package other applications/libraries as well). See our efforts at <https://salsa.debian.org/pan-team/soleil-packaging-overview>.

**Comment by ericpre** [comment]
@farhi, what is the process to update to latest release? Is there anything, we can do to help with updating?

**Comment by jlaehne** [comment]
https://github.com/hyperspy/hyperspy_gui_ipywidgets and https://github.com/hyperspy/hyperspy_gui_traitsui are not an independent user interface (which would be https://github.com/hyperspy/hyperspyUI), but contain the functionality necessary for pop-up/inline windows to adjust plots, settings, interact with certain functions (e.g. spikes or background removal). They are still helpful when working with Jupyter.

Both are also contained in the Arch User Repository (https://aur.archlinux.org/packages/?O=0&SeB=nd&K=hyperspy&outdated=&SB=n&SO=a&PP=50&do_Search=Go) and OpenSuse Community Repository (https://software.opensuse.org/search?utf8=%E2%9C%93&baseproject=ALL&q=hyperspy)

In the long run, other hyperspy extensions might be of interest as well: https://github.com/hyperspy/hyperspy-extensions-list

**Comment by picca** [comment]
Hello, the current situation is a non working hyperspy package in Debian unstalbe due to the imageio dependency.

the one available in Debian is 2.31.1.

Would it be possible to support this newer imageio library.

thanks for considering

**Comment by ericpre** [comment]
Thank you for the update. The IO code now lives in a separate library https://github.com/hyperspy/rosettasciio and the imageio requirement must have been sorted at some point - most likely in https://github.com/hyperspy/rosettasciio/pull/106.

If you are interested in the packaging of hyperspy in dedian, maybe you could help with the packaging of rosettasciio in debian as this would be needed for the hyperspy 2.0 release?

**Comment by picca** [comment]
> Thank you for the update. The IO code now lives in a separate library
> https://github.com/hyperspy/rosettasciio and the imageio requirement must have
> been sorted at some point - most likely in
> https://github.com/hyperspy/rosettasciio/pull/106.
> 
> If you are interested in the packaging of hyperspy in dedian, maybe you could
> help with the packaging of rosettasciio in debian as this would be needed for
> the hyperspy 2.0 release?

date of this hyperspy 2.0 release ?

This is just to be able to planify the upgrade in Debian

**Comment by ericpre** [comment]
It should be released by end of September!
